airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/parquet/JsonToAvroSchemaConverter.java

```diff
@@ -0,0 +1,211 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.destination.s3.parquet;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.base.Preconditions;
+import io.airbyte.commons.util.MoreIterators;
+import io.airbyte.integrations.base.JavaBaseConstants;
+import io.airbyte.integrations.destination.StandardNameTransformer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Type;
+import org.apache.avro.SchemaBuilder;
+import org.apache.avro.SchemaBuilder.RecordBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The main function of this class is to convert a JsonSchema to Avro schema. It can also
+ * standardize schema names, and keep track of a mapping from the original names to the standardized
+ * ones.
+ * <p/>
+ * For limitations of this converter, see the README of this connector:
+ * https://docs.airbyte.io/integrations/destinations/s3#parquet
+ */
+public class JsonToAvroSchemaConverter {
+
+  public static final Schema UUID_SCHEMA = LogicalTypes.uuid()
+      .addToSchema(Schema.create(Type.STRING));
+  private static final Logger LOGGER = LoggerFactory.getLogger(JsonToAvroSchemaConverter.class);
+  private static final Schema TIMESTAMP_MILLIS_SCHEMA = LogicalTypes.timestampMillis()
+      .addToSchema(Schema.create(Type.LONG));
+  private static final StandardNameTransformer NAME_TRANSFORMER = new StandardNameTransformer();
+
+  private final Map<String, String> standardizedNames = new HashMap<>();
+
+  static List<JsonSchemaType> getNonNullTypes(String fieldName, JsonNode typeProperty) {
+    return getTypes(fieldName, typeProperty).stream()
+        .filter(type -> type != JsonSchemaType.NULL).collect(Collectors.toList());
+  }
+
+  static List<JsonSchemaType> getTypes(String fieldName, JsonNode typeProperty) {
+    if (typeProperty == null) {
+      throw new IllegalStateException(String.format("Field %s has no type", fieldName));
```

Comment:
BTW some catalogs are producing fields without types, so it's not so uncommon... 

For example, the `source-facebook` does this I think... Does this exception cancel the sync of such catalogs? Should the fields be ignored or defaulted to a string for example instead?

Reply:
Got it. This is good to know.

In general, schemaless source is not suitable for Parquet. There is another big problem regarding how we are going to deal with `additionalProperties` whose types are known.

I will submit a follow-up PR to take care of it.

Reply:
Created an issue: https://github.com/airbytehq/airbyte/issues/4124